contrib/manyclangs-build-month

```diff
@@ -24,64 +24,209 @@ list_month() {
-
-    cd "$BUILD_DIR/elfshaker"
-    elfshaker pack "$YEAR$MONTH" --verbose
+_manyclangs_build_parts() {
+    [[ "${TRACE-}" ]] && set -x
+    set -euo pipefail # exported function, fresh shell.
+    YEAR=$1
+    MONTH=$2
+    shift 2
+
+    for part in "$@"
+    do
+        _manyclangs_build_part "$YEAR" "$MONTH" "$part"
+    done
 }
 
-main() {
-    MONTH="$1"
-    YEAR="$2"
-    date -d "$MONTH/01/$YEAR" 1>/dev/null || (print_help && exit 1)
+sanity_check() {
+    DEPS=(jq compdb2line ninja cmake "clang${BUILDER_CLANG_VERSION_SUFFIX}" "clang++${BUILDER_CLANG_VERSION_SUFFIX}")
+    MISSING_DEPS=()
+    for DEP in "${DEPS[@]}"
+    do
+        if ! command -v "$DEP" &> /dev/null;
+        then
+            MISSING_DEPS+=("$DEP")
+        fi
+    done
+
+    if [ ${#MISSING_DEPS[@]} -ne 0 ]
+    then
+        echo "Missing some binaries from \$PATH: ${MISSING_DEPS[*]}"
+        echo "Note that compdb2line is part of this project and needs installing."
+        exit 1
+    fi
 
-    BUILD_DIR="builds-$YEAR$MONTH"
-    if [ -d "$BUILD_DIR" ]; then
-        echo "$BUILD_DIR already exists! Aborting..."
+    if ! git rev-parse --git-dir &> /dev/null
+    then
+        echo 'Run with GIT_DIR set to path/to/llvm-project/.git or from inside a repository.'
         exit 1
     fi
+}
 
-    mkdir -p "$BUILD_DIR/elfshaker/elfshaker_data/packs"
-    echo "Builds will be done in $BUILD_DIR."
+progress_bar() {
+    COMMIT_COUNT=$1
+    pv --bytes --progress --fineta --timer --average-rate --line-mode --size "$COMMIT_COUNT" > /dev/null
+}
+
+apply_limit() {
+    if [ -z "${MANYCLANGS_LIMIT-}" ]; then cat; else head -n "$MANYCLANGS_LIMIT"; fi
+}
+
+apply_subset() {
+    if [ -z "${MANYCLANGS_SUBSET-}" ]; then cat; else awk 'NR%'"$MANYCLANGS_SUBSET"' == 0 { print }'; fi
+}
+
+main() {
+    sanity_check
+
+    echo
+
+    YEAR="$1"
+    MONTH="$2"
+    date -d "$YEAR/$MONTH/01" 1>/dev/null || { print_help; exit 1; }
+
+    BUILDER_CLANG_VERSION_SUFFIX=${BUILDER_CLANG_VERSION_SUFFIX--12}
+
+    commits_for_month "$YEAR" "$MONTH"  > manyclangs-commit-list.txt
+    apply_limit < manyclangs-commit-list.txt | apply_subset > manyclangs-commit-list.txt.subset
+    mv manyclangs-commit-list.txt.subset manyclangs-commit-list.txt
+    COMMIT_COUNT=$(wc -l manyclangs-commit-list.txt | awk '{print $1}')
+
+    NPARALLEL=${MANYCLANGS_NPARALLEL-5} # Number of parts to run in parallel.
+
+    if [[ "$NPARALLEL" == 1 ]] || command -v ninja-jobclient &> /dev/null
+    then
+        NPARTS=${MANYCLANGS_NPARTS-20} # Divide work into small equal-ish-sized parts.
+        split --number "l/$NPARTS" --numeric-suffixes manyclangs-commit-list.txt{,.}
+
+        export -f _manyclangs_build_parts _manyclangs_build_part _manyclangs_list_files0_to_pack
+        # A few extra slots, since we loose at least one to the build system and we
+        # don't expect every job to take 100% CPU.
+        JOBSLOP=5
+        command time -v \
+            jobserver -j$(($(nproc) + JOBSLOP)) \
```

Comment:
Which package installs `jobserver`? Can we mention that somewhere?

Reply:
Sadly there is no package, but it is available at https://github.com/olsner/jobclient

I agree this needs mentioning.